OsmAnd/src/net/osmand/plus/resources/ResourceManager.java

```diff
@@ -1194,16 +1196,31 @@ public List<Amenity> searchAmenities(SearchPoiTypeFilter filter, double topLatit
 					if (index != null && index.checkContainsInt(top31, left31, bottom31, right31)) {
 						List<Amenity> r = index.searchAmenities(top31,
 								left31, bottom31, right31, zoom, filter, matcher);
-						if (r != null) {
-							amenities.addAll(r);
-						}
+						sortedResult.put(cnt, r);
+						cnt++;
 					}
 				}
 			}
 		} finally {
 			searchAmenitiesInProgress = false;
 		}
-		return amenities;
+
+		// filter amenities for live-updates map
+		HashSet<Long> amenitiesHash = new HashSet<>();
+		List<Amenity> res = new ArrayList<>();
+		for (Map.Entry<Integer, List<Amenity>> entry : sortedResult.entrySet()) {
+			for (Amenity a : entry.getValue()) {
```

Comment:
>> sortedResult.entrySet()
Is nto sorted at all what the reason?

How is it different than old approach

Reply:
List<Amenity> r = index.searchAmenities(...) - synchronize method
amenities.addAll(r); - at the adding moment r is empty, so adding only link of r, but not r values(), values are adding randomly (depend from speed of *.obf reading).
Complete list amenities available only _try { }_ block
So I strictly set order by reading indexes.